File: code/216.combination-sum-iii.java
/*
 * @lc app=leetcode id=216 lang=java
 *
 * [216] Combination Sum III
 */

// @lc code=start

import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

class Solution {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> path = new LinkedList<>();
    public List<List<Integer>> combinationSum3(int k, int n) {
        backtracking(k, n, 1, 0);
        return res;
    }

    public void backtracking(int k, int n, int start, int sum) {
        if (sum > n) {
            return;
        }
        if (path.size() > k) {
            return;
        }

        if (sum == n && path.size() == k) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int i = start; i <= 9; i++) {
            path.add(i);
            backtracking(k, n, i + 1, sum + i);
            path.removeLast();
        }
    }
}
// @lc code=end

File: code/216.combination-sum-iii.go
/*
 * @lc app=leetcode id=216 lang=golang
 *
 * [216] Combination Sum III
 */

// @lc code=start
func combinationSum3(k int, n int) [][]int {
	path, res := make([]int, 0, k), make([][]int, 0)
	var backtracking func(k, n, sum, start int)
	backtracking = func(k, n, sum, start int) {
		if sum > n {
			return
		}

		if len(path) > k {
			return
		}

		if sum == n && len(path) == k {
			tmp := make([]int, k)
			copy(tmp, path)
			res = append(res, tmp)
			return
		}

		for i := start; i <= 9; i++ {
			path = append(path, i)
			backtracking(k, n, sum+i, i+1)
			path = path[:len(path)-1]
		}
	}
	backtracking(k, n, 0, 1)
	return res
}

// @lc code=end

